### DISCUSSION TITLE
Menu lag after loading a perf : is it normal ?

### DISCUSSION THREAD

**Discussion by Onosendaifr** [probonopd/MiniDexed#694] [Q&A] [answered]
When i load a performance, there is maybe 3s of lag before i'm able to press "Next" of "Previous" again, is it normal ? Is loading a perf take time ?

**Comment by Banana71** [comment]
Inputting with the encoder causes similar problems. After loading a performance, the display freezes for a short time, but the inputs on the encoder are then displayed with a delay. 
Loading the performance is very quick and you are ready to play straight away, but the display then goes to sleep for a short time 😴.

**Reply by Onosendaifr** [reply]
Strange ! Is it a bug ?

**Comment by Gitarrenzupfer** [comment]
Same behavior here with me. Raspberry 3 with I2C Display 2x16

**Reply by miotislucifugis** [reply]
I also have a pi3 and have not experienced this, tho' w/ a serial 1602 LCD, but i dont think i2c would be that much slower.   On mine, performances load in about 1/3rd of a second.    Perhaps try a different, better, faster sd card?

**Comment by Gitarrenzupfer** [comment]
I tried a new 32GB card (Sun Disk, a fast one). Same problem. It looks like something is preventing the GPIOs from receiving messages for about 3 seconds after the performance patch was changed. Maybe this behavior is a recent code change and your system is still running an older version? Mine is called "MiniDexed_2024-08-05-f6ebd78"

**Reply by miotislucifugis** [reply]
it may be worthwile to  double check that you have the gpio buttons wired correctly.   They should go low (to ground) when the switch is pressed.   I sorta remember having some weird lag when I had them mistakenly wired backwards (to 3.3v )  and the signal go would high w/ each press.   Minidexed still sorta worked, (I guess it was detecting the button release, not the press,)  but everything was slow and laggy.

oh,  also.... I assume that you meant "san" disk  not "sun" disk.    If the card actually says "sun" disk,  Id would definitely replace it and get a San Disk from a reliable vendor....(last part is important!)   There are a lot of fake cards out there. especially if you purchased from amazon , aliexpress, ebay, etc

**Comment by Banana71** [comment]
Since then I first noticed the behavior:

https://github.com/probonopd/MiniDexed/issues/583#issue-2029095590

**Comment by miotislucifugis** [comment]
So, is everyone that has this problem using I2C displays,  gadget mode,  or some other custom settings/hardware?   Just trying to find the common denominator.    Maybe list your hardware setup and any special changes to minidexed.ini that you have made to help narrow it down.

**Reply by Gitarrenzupfer** [reply]
I tried the minidexed.ini as it came without any changes exept 
1. sounddevice = I2s and 
2. my correct i2c address
no gadgetmode. The "Freeze" happens exactly at the same time as the [L] lights up in the display. After 2-3 seconds it works again. As in the short video with the 1306 OLED but a little bit longer

**Comment by Banana71** [comment]
Do you mean this problem?
my configuration:
RPi 4 , SSD1306, `` USBGadget=0 `` 

https://github.com/user-attachments/assets/c2b6c102-7eb1-4d91-a8f2-871c7ee3c1ef

**Reply by miotislucifugis** [reply]
Hey thanks for posting the video... that definitely helps clear up some of the confusion.... I thought that were were talking about navigating performance w/ gpios.    I just tested mine and the behavior is the same.   Im not sure that this is a bug;  just assumed that it was normal.   The delay is about a second or less, which seems reasonable as cpu loading 8 TG worth of info + all the other settings (effects, etc.)   The OP mentioned having a 3 second delay, which does sound excessive.   Maybe we can get a video of that ?

**Comment by ajboni** [comment]
https://github.com/user-attachments/assets/908ad49c-d302-4564-95e3-e51df3f2405f

Similar thing here rpi4

**Reply by Gitarrenzupfer** [reply]
exactly like my Pi3... 
Is your display connected via SPI or I2C ?

**Reply by ajboni** [reply]
its an LCD1602-I2C I2C display with a 'backpack'

<details><summary>Image</summary>
<p>
<img src="https://github.com/user-attachments/assets/82e21c0e-4a1e-419c-b5cc-31946a946c71"/>
</p>
</details>

**Comment by soyersoyer** [comment]
I experienced the same thing while developing #743, but with MIDI communication. After selecting a performance the display will refresh about ~50-100x with the same content. As a workaround, I added a "debouncer" to the MIDI queue to speed things up.

**Comment by Banana71** [comment]
If you are quick enough in your selection, something like this can happen:

https://github.com/user-attachments/assets/9ef93f4f-9de9-4dbb-9618-7ca9ce6476c2

Press, turn, load random Performance 😲🐒

**Comment by pik33** [comment]
I have the same behaviour on my RPi4 based Minidexed.

**Comment by diyelectromusic** [comment]
Just thinking about this issue once more - when this occurs, is the activity LED for the Pi also on?

It has to load performances from the SD card each time, which is probably the slowest thing MiniDexed does, so it may well be dependent on the performance of the SD card as to how long this delay is.

And it may be dependent on the number of files in the performance directory too.

I've wondered about loading the details in advance, but it isn't as easy as it is for voices due to the use of circles configuration files for performances...

Kevin

**Reply by soyersoyer** [reply]
When you load a performance, the display updates with each parameter change.
This helps:
https://github.com/soyersoyer/MiniDexed/commit/11276cca6022b0793c9c06072848644cf95604e4

**Reply by diyelectromusic** [reply]
Now included too.  Many thanks.

Kevin

**Comment by diyelectromusic** [comment]
Very experimental, but this might help, especially if running off slower SD cards with lots of performances...

https://github.com/probonopd/MiniDexed/pull/833

Feel free to give it a go if you're able to help test it.

Kevin

**Reply by diyelectromusic** [reply]
Ok, if the issue is less about the SD card, then this is a simpler fix that might also solve the problem... https://github.com/probonopd/MiniDexed/pull/834

Can you take a look?

Try this fix first, but if it doesn't solve the problem then try the other one and let me know :)

Thanks to @soyersoyer for this one.

Kevin

**Comment by probonopd** [comment]
#841 has an alternative solution for this, please test it. Thanks!

**Reply by Banana71** [reply]
Were there any problems with last week's solution? Unfortunately, the display is sluggish again in the current "Continuous Build."
I found the usability and behavior very good.

**Reply by probonopd** [reply]
The "Continuous Build" doesn't contain any of the improvements (neither last week's nor the one from https://github.com/probonopd/MiniDexed/pull/841). If you want to test #841, then you need to go to its page and scroll down for the latest "Build for testing". I hope it's not sluggish at all. Thanks!

**Comment by diyelectromusic** [comment, marked as answer]
This should now be addressed in the main branch thanks to https://github.com/probonopd/MiniDexed/pull/874

**Reply by probonopd** [reply]
Thanks @soyersoyer and @diyelectromusic
